computer_vision/connected_component_labeling.py

```diff
@@ -0,0 +1,242 @@
+        ...
+    ValueError: Matrix must be non-empty
+    >>> connected_component_labeling_4([[1, 1, 1], [1, 0], [1, 0, 1]])
+    Traceback (most recent call last):
+        ...
+    ValueError: Matrix must be rectangular
+    >>> connected_component_labeling_4([[1, 0, 0], [1, 0, 2], [1, 0, 1]])
+    Traceback (most recent call last):
+        ...
+    ValueError: Matrix must contain only 0s and 1s
+    """
+
+    if len(matrix) == 0 or len(matrix[0]) == 0:
+        raise ValueError("Matrix must be non-empty")
+    if not all(len(row) == len(matrix[0]) for row in matrix):
+        raise ValueError("Matrix must be rectangular")
+    if not all(all(x in (0, 1) for x in row) for row in matrix):
+        raise ValueError("Matrix must contain only 0s and 1s")
+
+    row_len = len(matrix[0])
+
+    # add padding
+    padded_matrix = (
+        [[0] * (row_len + 2)]
+        + [[0] + row + [0] for row in matrix]
+        + [[0] * (row_len + 2)]
+    )
+
+    # after the first traverse, we might label some elements incorrectly.
+    # when we spot incorrect labeling, we add neighbours label as a key
+    # and the element label as a value
+    mapping: dict = {}
+
+    next_label = 1
+
+    # first traverse
+    for i in range(1, len(padded_matrix) - 1):
+        for j in range(1, row_len + 1):
+
+            if padded_matrix[i][j] == 1:
+
+                neighbours = [padded_matrix[i - 1][j], padded_matrix[i][j - 1]]
+
+                if any(neighbours):
+
+                    min_label = min([v for v in neighbours if v > 0])
+                    padded_matrix[i][j] = min_label
+
+                    for label in neighbours:
+                        # if any of the neighbors appears to have label \
+                        # greater than current element - add it's \
+                        # label to mapping
+                        if label > min_label:
+                            if label not in mapping:
+                                mapping[label] = set()
+                            mapping[label].add(min_label)
+
+                else:
+                    padded_matrix[i][j] = next_label
+                    next_label += 1
+
+    # "unpad" the matrix
+    matrix = [row[1:-1] for row in padded_matrix[1:-1]]
+
+    # restructure the mapping, so that:
+    # 3: (2, 1)         3: (2, 1)
+    # 6: (3)        ->  6: (3, 1)
+    #                   2: (1,)
+    for key in sorted(mapping.keys()):
+        adj_labels = mapping[key]
+        min_label = min(adj_labels)
+        for adj_label in adj_labels:
+            if adj_label not in mapping:
+                mapping[adj_label] = set()
+            mapping[adj_label].add(min_label)
+
+    # apply the mapping
+    for i in range(len(matrix)):
+        for j in range(row_len):
+            if matrix[i][j] in mapping:
+                matrix[i][j] = min(mapping[matrix[i][j]])
+
+    return matrix
+
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
```

Comment:
```suggestion
    doctest.testmod()

```
Files should end in a new line

Reply:
As far as I'm concerned, github automatically remove one space in the browser from the end of the file.

If the space was not actually at the end of the file, github would have been marked it as shown below:
![image](https://user-images.githubusercontent.com/32400447/195177146-df76edf2-d22e-4834-b74e-58fe9e6a7946.png)

You can check my fork, I removed the space from the end of the file, and added it to the fork repository.

Thank you :)

Reply:
I accidentally added a file to the pull-request, could you help me remove it?